### PR TITLE
Add a `fmt` subcommand

### PR DESCRIPTION
**Pull request by liamdawson**
IAMy is rather rigid about the YAML format, which has tripped me up multiple times in drift detection scenarios. For example, if I applied this policy:

```yaml
Policy:
  Statement:
    - Sid: ExampleStatement  
      Action:
        - ec2:TerminateInstances
        - ec2:StopInstances
        - 'iam:*' # needed for reason
      Effect: Allow
      Resource:
        - '*'
  Version: 2012-10-17
```

it would be replaced with this during the next `iamy pull`:

```yaml
Policy:
  Statement:
  - Action:
    - ec2:StopInstances
    - ec2:TerminateInstances
    - iam:*
    Effect: Allow
    Resource: '*'
    Sid: ExampleStatement
  Version: "2012-10-17"
```

That is, each of the following things caused changes:

- Keys and statement actions out of alphabetical order
- Passing a single resource as an array of one element
- Unquoted version value
- Spaces before the `-` for array items
- Trailing spaces on lines
- Comments

All of these can be corrected by reading the YAML files and writing them back in a "normalized" fashion. I believe this would be useful as a CI check.

### Progress

- [x] Add a normalize command
- [x] Add tests
- [ ] Make dry-run work?

### CODE REVIEW

File: normalize_test.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+	return nil
+}
+
+func newUi() Ui {
+	return Ui{
+		Logger: log.New(ioutil.Discard, "", 0),
+		Error:  log.New(ioutil.Discard, "", 0),
+		Debug:  log.New(ioutil.Discard, "", 0),
+		Exit:   os.Exit, // TODO: probably wrong
```
Comment (patrobinson) [inline normalize_test.go]: It doesn't look like this is even called from anywhere so 🤷

File: iamy.go [hunk 2]
```diff
@@ -40,16 +40,19 @@ const cloudformationStackNameTag = "aws:cloudformation:stack-name"
...
+		pull               = kingpin.Command("pull", "Syncs IAM users, groups and policies from the active AWS account to files")
+		pullDir            = pull.Flag("dir", "The directory to dump yaml files to").Default(defaultDir).Short('d').String()
+		pullCanDelete      = pull.Flag("delete", "Delete extraneous files from destination dir").Bool()
+		lookupCfn          = pull.Flag("accurate-cfn", "Fetch all known resource names from cloudformation to get exact filtering").Bool()
+		push               = kingpin.Command("push", "Syncs IAM users, groups and policies from files to the active AWS account")
+		pushDir            = push.Flag("dir", "The directory to load yaml files from").Default(defaultDir).Short('d').ExistingDir()
+		normalize          = kingpin.Command("normalize", "Normalize YAML files to reduce potential drift")
+		normalizeDir       = normalize.Flag("dir", "The directory to normalize yaml files within").Default(defaultDir).Short('d').ExistingDir()
+		normalizeCanDelete = normalize.Flag("delete", "Delete extraneous files from destination dir").Bool()
```
Comment (patrobinson) [inline iamy.go]: What does this default to?

**Review by patrobinson — DISMISSED** [review]
Ideally I'd like to normalise data from files and API so that what iamy has in memory before it does a comparison is consistent. This should be using some form of structure instead of yaml.

This prevents the issues this resolves from happening at all. As long as they are "effectively equivalent" then we don't need a normalise command at all.

**Comment by liamdawson** [comment]
> Ideally I'd like to normalise data from files and API so that what iamy has in memory before it does a comparison is consistent. This should be using some form of structure instead of yaml.

I'd love to see that at some point, but I imagine it will take a while. Do you feel this feature is a useful helper in the interim?

File: iamy.go (hunk 2 above)
Comment (liamdawson) [reply iamy.go]: Defaults to `false`

**Comment by patrobinson** [comment]
@liamdawson yes I think so

File: iamy.go [hunk 3]
```diff
@@ -40,16 +40,19 @@ const cloudformationStackNameTag = "aws:cloudformation:stack-name"
...
-		push          = kingpin.Command("push", "Syncs IAM users, groups and policies from files to the active AWS account")
-		pushDir       = push.Flag("dir", "The directory to load yaml files from").Default(defaultDir).Short('d').ExistingDir()
+		debug              = kingpin.Flag("debug", "Show debugging output").Bool()
+		skipCfnTagged      = kingpin.Flag("skip-cfn-tagged", fmt.Sprintf("Shorthand for --skip-tagged %s", cloudformationStackNameTag)).Bool()
+		skipTagged         = kingpin.Flag("skip-tagged", "Skips IAM entities (or buckets associated with bucket policies) tagged with a given tag").Strings()
+		includeTagged      = kingpin.Flag("include-tagged", "Includes IAM entities (or buckets associated with bucket policies) tagged with a given tag").Strings()
+		pull               = kingpin.Command("pull", "Syncs IAM users, groups and policies from the active AWS account to files")
+		pullDir            = pull.Flag("dir", "The directory to dump yaml files to").Default(defaultDir).Short('d').String()
+		pullCanDelete      = pull.Flag("delete", "Delete extraneous files from destination dir").Bool()
```
Comment (andrewjhumphrey) [inline iamy.go]: I wonder whether we should offer the --delete flag for the normalise command, I can see someone wanting to do a "full" cleanup.

**Comment by andrewjhumphrey** [comment]
Love the feature, but why the 'z' :smile:

File: iamy.go (hunk 3 above)
Comment (liamdawson) [reply iamy.go]: We do: line 55

Comment (andrewjhumphrey) [reply iamy.go]: Ahh, I should read more

Comment (andrewjhumphrey) [reply iamy.go]: I'm torn as to whether to duplicate it like this, or to push it up to kingpin and tolerate the ability to uselessly supply it to the push command.

Comment (liamdawson) [reply iamy.go]: I used `pushDir`/`pullDir` as the precedent, noting that the fact that `dryRun` is global meant I had to write a failure case to indicate that it's not supported. If I did anything more here, I'd:

1. Hoist `dir` to kingpin
2. Lower `dryRun` to `push`/`pull`, and
3. Leave `delete` where it is

I'm personally quite fussy about CLI behaviour, but I'll happily defer to the ideal choices for the application.

**Comment by liamdawson** [comment]
> Love the feature, but why the 'z' 😄

Actually, I wonder whether something like `iamy rewrite` might be a better name (and avoid your locale woes)

**Review by patrobinson — APPROVED** [review]

**Review by andrewjhumphrey — APPROVED** [review]
Happy to go with this as is,  I note that you had a check box in the description for "make dry-run work?"  do you want to sort that out in this PR before merging?  Or push that to future work?

**Comment by liamdawson** [comment]
> Happy to go with this as is, I note that you had a check box in the description for "make dry-run work?" do you want to sort that out in this PR before merging? Or push that to future work?

Ship it as-is 👍

**Pull request merged**